File: 多边形拟合.py
import cv2
import numpy as np

img = cv2.imread("kuang.PNG")
#创建蓝色掩膜（HSV）
hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
lower_blue = np.array([100, 150, 100])
upper_blue = np.array([140, 255, 255])
#二值纯膜生成
blue_mask = cv2.inRange(hsv, lower_blue, upper_blue)
#形态学连接
kernel_large = np.ones((20, 20), np.uint8)
blue_mask = cv2.morphologyEx(blue_mask, cv2.MORPH_CLOSE, kernel_large)
#去除噪点（腐蚀和膨胀）
kernel_small = np.ones((3, 3), np.uint8)
blue_mask = cv2.morphologyEx(blue_mask, cv2.MORPH_OPEN, kernel_small)
#找轮廓
contours, hierarchy = cv2.findContours(blue_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
#副本
result = img.copy()
#轮廓与面积处理
for i, contour in enumerate(contours):
    area = cv2.contourArea(contour)
    if area < 1000:
        continue
    x, y, w, h = cv2.boundingRect(contour)
    aspect_ratio = max(w, h) / min(w, h)
    cv2.rectangle(result, (x, y), (x + w, y + h), (0, 0, 255), 2)

cv2.imshow("first", blue_mask)
cv2.imshow("second", result)
cv2.waitKey(-1)
cv2.destroyAllWindows()

File: 多边形拟合.cpp
#include <opencv2/opencv.hpp>
#include<vector>
int main() {

    cv::Mat img = cv::imread("C:/Users/16968/Desktop/123/1280x1280.PNG");

    cv::Mat hsv;
    cv::cvtColor(img, hsv, cv::COLOR_BGR2HSV);


    cv::Scalar low_blue(95,120, 70);
    cv::Scalar high_blue(110, 255, 255);

    // 创建蓝色掩膜
    cv::Mat blue_mask;
    cv::inRange(hsv, low_blue, high_blue, blue_mask);
    cv::medianBlur(blue_mask,blue_mask,3);
    cv::bitwise_not(blue_mask,blue_mask);

    cv::Mat blue_result;
    cv::bitwise_and(img, img,blue_result, blue_mask);
    std::vector<std::vector<cv::Point> > cont;
    std::vector<cv::Vec4i> l;
    cv::findContours(blue_mask,cont,l,cv::RETR_LIST,cv::CHAIN_APPROX_NONE);
    cv::Mat first=img.clone();
    cv::drawContours(first,cont,-1,(0,0,255),3);
    cv::Mat result=img.clone();
    for (int i=0;i<cont.size();i++) {
        double area=cv::contourArea(cont[i]);
        if (area>250 && area<3000) {
            // 正确计算 epsilon
            double epsilon = 0.03 * cv::arcLength(cont[i], true);

            // 正确使用 approxPolyDP
            std::vector<cv::Point> approx;
            cv::approxPolyDP(cont[i], approx, epsilon, true);

            if (approx.size() <= 6) {
                cv::Rect rect = cv::boundingRect(cont[i]);
                cv::rectangle(result, rect, cv::Scalar(0, 0, 255), 2);
            }
            cv::Rect rect=cv::boundingRect(cont[i]);
            cv::rectangle(result,rect,cv::Scalar(0,0,255),3);
        }
    }

    cv::imshow("contours", first);
    cv::imshow("result", result);

    cv::waitKey();
    cv::destroyAllWindows();

    return 0;
}